### PR TITLE
Fix ordering of max in flight check

### LINKED ISSUE #5452: Builds can be stuck in max in flight with a mix of rerun and regular builds
## Summary

Seems like the fix with #5331 missed a part where the builds that are fetching during the next pending build for max in flight query 

https://github.com/concourse/concourse/blob/fab3de1722a2ce998d3710bd066453594f24ec57/atc/db/job.go#L937-L948

should also be ordered in the same way that grabbing the next pending builds are ordered.

https://github.com/concourse/concourse/blob/fab3de1722a2ce998d3710bd066453594f24ec57/atc/db/job.go#L625-L632

## Steps to reproduce

Have two builds, one rerun build and one regular build, where the rerun build is created before the regular build. They will both be stuck with max in flight reached.

## Expected results

Max in flight should not be reached.

### PR DESCRIPTION
**Pull request by clarafu**
# Existing Issue

Fixes #5452.

# Why do we need this PR?
There is a bug where if you create a new pending build and then trigger a rerun pending build, all the builds will be stuck in pending state. This is because the ordering of the builds that are scheduled are different than the ordering of the builds that are used to check if the job has reached max in flight. The pending builds are scheduled in the same order as the algorithm where it takes into account rerun builds and orders them in the order of the build it is a rerun of. For example, if you have three pending builds that were created in this order, 5, 6.1, 2.1, where 6.1 and 2.1 are reruns, we will schedule them in the order 2.1, 5, 6.1. This is because we schedule builds from oldest -> latest and since 2.1 is a rerun of 2, it gets scheduled first. Then 5 is scheduled because it is older than the build 6.1 was a rerun of. This is the part of the query that takes care of this ordering.

https://github.com/concourse/concourse/blob/3d7c18cc7d7efeabd6f3b062ddc662aa39c4bc88/atc/db/job.go#L630

With the scheduler ordering the pending builds that it needs to schedule in this format, the max in flight check currently orders them just by build ID in ascending format. 

https://github.com/concourse/concourse/blob/3d7c18cc7d7efeabd6f3b062ddc662aa39c4bc88/atc/db/job.go#L945

With the max in flight check ordering it in build ID, the same example as before with pending builds 5, 6.1 and 2.1 created in this order, the max in flight check would have determined build 5 to be the next pending build. But the next pending build according to the scheduler is 2.1 so then it compares if the next pending build according to the scheduler (build 2.1) to the next pending build according to the max in flight check (build 5). Since they are different, it returns max in flight reached.

https://github.com/concourse/concourse/blob/3d7c18cc7d7efeabd6f3b062ddc662aa39c4bc88/atc/db/job.go#L844

# Changes proposed in this pull request
I made a change that orders the max in flight check for the next pending build the same way as the scheduler orders it's pending build. This will ensure that both definitions of what the "next pending build" is will be the same.

I didn't add any tests because this is an edge case that can't be properly written as a unit test since it involves certain conditions on multiple components. However, I did do manual testing with one job and creating a new pending build first and then a rerun build. I observed it be stuck in pending state before this change and after the change, is able to start.

# Contributor Checklist
> Are the following items included as part of this PR? Please delete checkbox items that don't apply.
- [ ] Unit tests
- [ ] Integration tests (if applicable)
- [ ] Updated documentation (located at https://github.com/concourse/docs)
- [ ] Updated release notes (located at https://github.com/concourse/concourse/tree/master/release-notes)

# Reviewer Checklist
> This section is intended for the core maintainers only, to track review progress. Please do not
> fill out this section.
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the [BOSH](https://github.com/concourse/concourse-bosh-release) 
      and [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for the [integration tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env) (for example, if they are Garden configs that are not displayed in the `--help` text).

### CODE REVIEW

**Comment by jlamb1** [comment]
any update on this issue? We're currently compiling a custom binary to get builds running again. thank you!

**Comment by clarafu** [comment]
@jlamb1 Just waiting on the testflight test to pass and a review then it should be ready to be in the next release!

**Comment by jlamb1** [comment]
leaving this here in case it helps someone else. Within the database running:

`update builds set status 'succeeded' where rerun_number is not NULL and status='pending'`

Will clear out the hung jobs and allow new jobs to commence. This is *not* a fix, but short term help.

File: atc/db/job.go [hunk 1]
```diff
@@ -934,15 +934,15 @@ func (j *job) getRunningBuildsBySerialGroup(tx Tx, serialGroups []string) ([]Bui
 }
 
 func (j *job) getNextPendingBuildBySerialGroup(tx Tx, serialGroups []string) (Build, bool, error) {
-	row := buildsQuery.Options(`DISTINCT ON (b.id)`).
+	row := buildsQuery.
```
[the remark this answers is not in the document]
Comment (xtreme-sameer-vohra) [reply atc/db/job.go:937]: Given the `buildsQuery`, I can't see the need for `Distinct`

**Comment by xtreme-sameer-vohra** [comment]
There is another instance that could be cleaned up https://github.com/concourse/concourse/blob/master/atc/db/job.go#L906

File: atc/db/job.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (clarafu) [reply atc/db/job.go:937]: @xtreme-sameer-vohra Ya I don't think the distinct was necessary since we are just querying from the builds table. I tested the query out without the distinct and it seems to work properly!

[the remark this answers is not in the document]
Comment (clarafu) [reply atc/db/job.go:937]: Oh now that I look at it again, the distinct might be for if there are multiple rows for a job from the `jobs_serial_groups` table that it joins with. I originally thought it shouldn't need the distinct but I forgot about the fact that jobs can be in multiple serial groups. Thanks for pointing this out!! I will add back the distinct!

[the remark this answers is not in the document]
Comment (xtreme-sameer-vohra) [reply atc/db/job.go:937]: ah, I didn't realize, the `Options` param would apply to the join on `jobs_serial_groups` as well

[the remark this answers is not in the document]
Comment (clarafu) [reply atc/db/job.go]: I think it's because if a job is in multiple serial groups, it will have multiple rows in the `jobs_serial_groups` table and when we query for a build from that job, it will join to the serial groups table combining the build row with all the rows for that job in the `jobs_serial_groups` table (resulting in multiple rows for the same build).

File: atc/db/job.go [hunk 2]
```diff
@@ -934,21 +934,26 @@ func (j *job) getRunningBuildsBySerialGroup(tx Tx, serialGroups []string) ([]Bui
...
-		RunWith(tx).
-		QueryRow()
+		ToSql()
+	if err != nil {
+		return nil, false, err
+	}
+
+	row := tx.QueryRow(`
+			SELECT * FROM (`+subQuery+`) j
+			ORDER BY COALESCE(rerun_of, id) ASC, id ASC
```
Comment (xtreme-sameer-vohra) [inline atc/db/job.go:952]: Not sure if this would be an issue based on the[ postgres docs](https://www.postgresql.org/docs/9.0/sql-select.html#SQL-DISTINCT), which mention that `Distinct on` should be executed with a matching `order by` as the select query might have unpredictable results otherwise.

We used this to test that out and received different results for the flat and nested queries

```
CREATE TABLE test (
    id SERIAL PRIMARY KEY,
    name text NOT NULL,
    location text NOT NULL,
    baz text NOT NULL
)
```

```
insert into test (name,location, baz) values ('foo', 'nyc', 'bar');
insert into test (name,location, baz) values ('foo', 'nyc', 'bar');
insert into test (name,location, baz) values ('foo', 'tor', 'bar');
insert into test (name,location, baz) values ('foo', 'la', 'bar');
insert into test (name,location, baz) values ('moo', 'nyc', 'bar');
insert into test (name,location, baz) values ('moo', 'tor', 'bar');
insert into test (name,location, baz) values ('moo', 'la', 'bar');
insert into test (name,location, baz) values ('soo', 'la', 'bar');
```

flat query
```
SELECT DISTINCT ON (name) name, location
    FROM test
    ORDER BY name, location DESC;
```


nested query
```
select * from (SELECT DISTINCT ON (name) name, location
    FROM test) j order by name, location DESC;
```

Comment (xtreme-sameer-vohra) [reply atc/db/job.go:952]: chatted about this, since only the build is scanned from the result, the `distinct on` behaviour shouldn't impact the build.

**Comment by clarafu** [comment]
@xtremerui @xtreme-sameer-vohra There is a test here https://github.com/concourse/concourse/blob/master/atc/db/job_test.go#L1171 that tests for a job that belongs to multiple serial groups. I'm not sure if you think that is enough? Maybe we can merge this for now and then I can add another test case if you think it isn't enough to cover this.
